File: chap-1/1-3.cpp
#include <iostream>
/*
 * Design algorithm to remove the duplicate characters
 * in a string without using an additional buffer. 
 * Can use additional buffer
 */


void removeDuplicates(char *str);

int main(int argc, char *argv[]){

	char str[] = "hello";
	char str2[] = "racecar";

	removeDuplicates(str);
	removeDuplicates(str2);

	return 0;
}

void removeDuplicates(char *str){

	/*
	 * Solution:
	 * 	hello
	 * 	h   
	 * 	e
	 * 	l dup i=3
	 *
	 * 	1: For each character, compare it to all preceding characters
	 * 	If that character is in the string, skip it for adding to the 
	 * 	resultant string
	 * 	O(n^2) - no memory buffer needed
	 *
	 *
	 * 	2: Buffer used
	 *
	 * 	orig-buf -> racecar
	 *
	 *	secd-buf -> r:0, a:1, c:2, e:3, c, a, r
	 *
	 *	For each character add it to dictionary if it doesn't exist
	 *	key is character, value is index. dict is not necessarily sorted by value
	 *	sorting would require O(n logn)	
	 *
	 *
	 * 	3: Buffer 2 
	 *
	 * 	orig-buf -> helloworld
	 *
	 * 	keep track of seen characters
	 * 	Use array instead of dictionary
	 *
	 * 	What kind of char set? if ascii, then max 256 array for array buffer
	 * 	Each index corresponds to letter
	 *
	 * 	What is value at index?
	 * 		Boolean
	 * 		Index of character in string
	 *
	 * 	If boolean, then we can use the boolean array to say if we've seen this character
	 * 	then we can use a similar tail method from the first solution 
	 *
	 * 	- Iterate over the string
	 * 	- once we see a character, set the value of the array at the 
	 * 	- character index to be true
	 *
	 * 	While iterating, if we have seen the current character skip it
	 * 	if we havent seen it, update the end of the string 
	 *
	 *
	 *
	 * 	
	 */
	
	int len = strlen(str);
	if (str == NULL){
		return;
	}
	if (len < 2) {
		return;
	}

	int tail = 0;
	bool char_set[256] = {false};

	std::cout << "Original: " << str << std::endl;

	// Start from the second character 
	for(int i = 0; i < len; i++){
		// Check if we've seen it
		if(!char_set[str[i]]){
			char_set[str[i]] = true;
			str[tail++] = str[i];
		}
	}
	str[tail] = '\0';

	std::cout << "Deduplicated: " << str << std::endl;
}

File: chap-1/python/1-3.py
#!/usr/bin/env python
"""
Problem: design algorithm and write code to remove duplicate characters in a string
         without an extra buffer (one or two variables okay)

Solution:
        1. First detect duplicate characters
        2. Remove them in place (i.e. shift characters once done)

        Could possibly store indices of duplicates
        For each index move the character at index+1 over to left
        Special case (end of string) just collapse string don't look at next character

        "Hello World"

        If we want to do O(n), we need some sort of structure to store duplicates like a hash
        Or we can use bit vector

        current dup = l
        current index = 3

        what kind of variables (array, or just char/integer)


"""

def removeDuplicates(word):
    checker = 0
    num_shifts = 0
    print word
    for i in range(0, len(word)):
        i -= num_shifts
        num_shifts = 0
        if i >= len(word):
            break
        if word[i] == ' ':
            continue
        val = ord(word[i]) - ord('a')
        if (checker & (1 << val)) > 0:
            if i + 1 < len(word):
                word = word[:i] + word[i+1:] 
                num_shifts = 1
        checker |= (1 << val)

    print word


def main():
    removeDuplicates("hello")
    removeDuplicates("helloworld")

if __name__ == "__main__":
    main()

